cirq-core/cirq/value/condition.py

```diff
@@ -0,0 +1,168 @@
+        return KeyCondition(keys[0])
+
+    def __str__(self):
+        return str(self.key)
+
+    def resolve(self, measurements: Mapping[str, Sequence[int]]) -> bool:
+        key = str(self.key)
+        if key not in measurements:
+            raise ValueError(f'Measurement key {key} missing when testing classical control')
+        return any(measurements[key])
+
+    def _json_dict_(self):
+        return json_serialization.dataclass_json_dict(self)
+
+    @property
+    def qasm(self):
+        return f'm_{self.key}!=0'
+
+
+@dataclasses.dataclass(frozen=True)
+class SympyCondition(Condition):
+    """A classical control condition based on a sympy expression.
+
+    This condition resolves to True iff the sympy expression resolves to a
+    Truthy value when the measurement keys are substituted in as the free
+    variables.
+
+    To account for the fact that measurement key strings can contain characters
+    not allowed in sympy variables, we use x0..xN for the free variables and
+    substitute the measurement keys in by order. For instance a condition with
+    `expr='x0 > x1', keys=['0:A', '0:B']` would resolve to True iff key `0:A`
+    is greater than key `0:B`.
+
+    The `cirq.parse_sympy_condition` function automates setting this up
+    correctly. To create the above expression, one would call
+    `cirq.parse_sympy_condition('{0:A} > {0:B}')`.
+    """
+
+    expr: sympy.Expr
+    control_keys: Tuple['cirq.MeasurementKey', ...]
+
+    @property
+    def keys(self):
+        return self.control_keys
+
+    def with_keys(self, *keys: 'cirq.MeasurementKey'):
+        assert len(keys) == len(self.control_keys)
+        return dataclasses.replace(self, control_keys=keys)
+
+    def __str__(self):
+        replacements = {f'x{i}': f'{{{str(key)}}}' for i, key in enumerate(self.control_keys)}
+
+        class CustomCodePrinter(sympy.printing.StrPrinter):
+            def _print_Symbol(self, expr):
+                return replacements[expr.name]
+
+        return CustomCodePrinter().doprint(self.expr)
+
+    def resolve(self, measurements: Mapping[str, Sequence[int]]) -> bool:
+        missing = [str(k) for k in self.keys if str(k) not in measurements]
+        if missing:
+            raise ValueError(f'Measurement keys {missing} missing when testing classical control')
+
+        def value(k):
+            return digits.big_endian_bits_to_int(measurements[str(k)])
+
+        replacements = {f'x{i}': value(k) for i, k in enumerate(self.keys)}
+        return bool(self.expr.subs(replacements))
+
+    def _json_dict_(self):
+        return json_serialization.dataclass_json_dict(self)
+
+    @property
+    def qasm(self):
+        raise NotImplementedError()
+
+
+def parse_sympy_condition(s: str) -> 'cirq.SympyCondition':
+    """Parses a string into a `cirq.SympyCondition`.
+
+    The measurement keys in a sympy condition string must be wrapped in curly
+    braces to denote them. For example, to create an expression that checks if
```

Comment:
Potentially awkward if someone is using curly braces in their measurement key names...but I suppose they could always fall back on the base `SympyCondition` constructor if that's the case. We should include a comment here that reflects this.

Reply:
I was thinking about this, and shouldn't we limit the characters allowed in measurement keys at some point, even if it's breaking? This isn't the first time we've needed to deal with arbitrary characters in the names and I don't think it'll be the last either. (I imagine `[` will need special handling when we introduce repeated measurements).

Reply:
There's some precedent for disallowing characters in measurement keys:
https://github.com/quantumlib/Cirq/blob/11457220ab77e1e41d58b8ab3b2a98833d77d03d/cirq-core/cirq/value/measurement_key.py#L46-L47
That said, it's technically a breaking change, and we should avoid limiting it too much. Could you open an issue to make sure we address this before Cirq 1.0?

Reply:
We allow spaces too. I can see that causing confusion because we can't trim the inside of the braces here. `expr='{ a } < { b }'` must bind to keys space-a-space and space-b-space.

Reply:
I added `\` as an escape character and documented it and added some tests. So now users can put whatever characters they want in their keys.

I ended up having to change the parsing code to go character-by-character. IDK if there's a good way of doing it with regex (I have zero regex foo).